File: CodingTest/Graph/BFS/13913_숨바꼭질4.py
from collections import deque
import sys
sys.stdin = open("input.txt", "r")

n,k = map(int, input().split())

visited = [-1]*100001
path = [] # 지금까지 방문 경로 저장
def bfs(start, target):
    queue = deque() 
    queue.append((n,0))
    visited[start] = start
    while queue:
        cur, cur_time = queue.popleft()        
        if cur == target:  # 동생을 찾음
            idx = cur
            while idx != start:
                path.append(idx)
                idx = visited[idx] # 이 부분을 잘 모르겠다.
            path.append(start)
            return cur_time
        if cur+1 < 100001 and visited[cur+1] == -1 :
            queue.append((cur+1,cur_time+1))
            visited[cur+1] = cur
             
        if cur-1 >= 0 and visited[cur-1] == -1 :
            queue.append((cur-1,cur_time+1))
            visited[cur-1] = cur
            
        if cur*2 < 100001 and visited[cur*2] == -1 :
            queue.append((cur*2,cur_time+1))
            visited[cur*2] = cur


print(bfs(n,k))
print(*path[::-1]) # 뒤에서 부터 출력

File: CodingTest/Graph/BFS/13913_숨바꼭질4.cpp
#include<bits/stdc++.h>
#define endl "\n"
#define MAX 100000+1
using namespace std;

int arr[MAX];
bool visited[MAX];
int move_time[3] = {1,1,1}; // x*2, x-1, x+1,  로 이동하는데 걸리는 시간
int answer = 0;
int n,k;
void bfs(int start)
{
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>>pq;    
    pq.push({0,start});// 걸린 시간, 현재 위치
    arr[start] = 0;
    visited[start] = true;
    while (!pq.empty())
    {
        int t = pq.top().first;
        int pos = pq.top().second;
        pq.pop();
        if(pos==k){
            answer = t;
            return;
        }
        for(int i=0; i<3; i++){
            int next_time = t + move_time[i];
            int next_pos=0;
            if(i==0) next_pos = pos*2;
            if(i==1) next_pos = pos-1;
            if(i==2) next_pos = pos+1;
            if(next_pos<0 || next_pos >MAX || arr[next_pos]!=-1)continue; // 범위 초과 및 재방문 확인
            if(next_pos==k){
                answer = next_time;
                return;
            }
            pq.push({next_time,next_pos});
            arr[next_pos] = next_time;
            cout<<"pos:"<<next_pos<<" time:"<<next_time<<endl;
        }
    }
    
}
int main() {	
	ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
	freopen("input.txt", "r", stdin);
	cin>>n>>k;
    for(int i=0; i<MAX; i++){
        arr[i]=-1;
    }
    bfs(n);
    cout<<answer;
	return 0;
}